SYCL/ESIMD/regression/copyto_char_test.cpp

```diff
@@ -0,0 +1,160 @@
+// REQUIRES: gpu
+// UNSUPPORTED: cuda || hip
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+
+//==- copyto_char_test.cpp - Test for using copy_to to copy char buffers -==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <sycl/ext/intel/esimd.hpp>
+#include <sycl/sycl.hpp>
+
+#include <iostream>
+#include <memory>
+
+using namespace sycl::ext::intel::esimd;
+
+template <typename DataT>
+using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+template <typename DataT>
+using shared_vector = std::vector<DataT, shared_allocator<DataT>>;
+
+using DataT = char;
+
+template <int NumElems, bool IsAcc> int test_to_copy() {
+  sycl::queue queue;
+  constexpr int NumSelectedElems = NumElems / 3;
+  constexpr int Stride = 2;
+  constexpr int Offset = 6;
+
+  shared_allocator<DataT> allocator(queue);
+  shared_vector<DataT> result(NumElems, allocator);
+  shared_vector<DataT> initial_ref_data(NumElems, allocator);
+  shared_vector<DataT> ref_data_for_fill(NumSelectedElems, allocator);
+
+  constexpr size_t value_for_increase_ref_data_for_fill = 50;
+  for (size_t i = 0; i < NumElems; ++i) {
+    initial_ref_data[i] = i + 1;
+  }
+  // We should have different values in the first reference data and in the
+  // second reference data.
+  for (size_t i = 0; i < NumSelectedElems; ++i) {
+    ref_data_for_fill[i] =
+        initial_ref_data[i] + value_for_increase_ref_data_for_fill;
+  }
+  if constexpr (IsAcc) {
+    sycl::buffer<DataT> output_buf(result.data(), result.size());
+    queue.submit([&](sycl::handler &cgh) {
+      DataT *init_ref_ptr = initial_ref_data.data();
+      DataT *ref_data_for_fill_ptr = ref_data_for_fill.data();
+      auto acc =
+          output_buf.template get_access<sycl::access::mode::read_write>(cgh);
+
+      cgh.single_task([=]() SYCL_ESIMD_KERNEL {
+        simd<DataT, NumElems> src_simd_obj;
+        src_simd_obj.copy_from(init_ref_ptr);
+        simd<DataT, NumSelectedElems> dst_simd_obj;
+        dst_simd_obj.copy_from(ref_data_for_fill_ptr);
+        src_simd_obj.template select<NumSelectedElems, Stride>(Offset) =
+            dst_simd_obj;
+        src_simd_obj.copy_to(acc, 0);
+      });
+    });
+  } else {
+    queue.submit([&](sycl::handler &cgh) {
+      DataT *init_ref_ptr = initial_ref_data.data();
+      DataT *ref_data_for_fill_ptr = ref_data_for_fill.data();
+      DataT *const out_ptr = result.data();

```

Comment:
Most likely usm allocator returns well aligned memory (may be 16-byte aligned).
The test could be more strong if it passed unaligned memory.
Something like: 'out_ptr = result.data() + 1;', which would require result to be a little bigger and minor changes in the results-checkin code.
I guess, if you pass unaligned memory here, then the proposed fix inside implementation of copy_to() may fail as it uses block_store.

Reply:
Thank you.